### PR TITLE
Authentication in SPA

### LINKED ISSUE #58: Make SPA send token to API

### PR DESCRIPTION
**Pull request by laska-sl**

### CODE REVIEW

File: ArcadiaParties-SPA/src/app/app.component.ts [hunk 1]
```diff
@@ -1,10 +1,30 @@
...
   templateUrl: './app.component.html',
   styleUrls: ['./app.component.scss']
 })
 export class AppComponent {
-  constructor() {}
+  str: string;
+  constructor(private http: HttpClient, private authService: AuthService) { }
+
+
+  login() {
```
Comment (Duveshka) [inline ArcadiaParties-SPA/src/app/app.component.ts]: Empty line

File: ArcadiaParties-SPA/src/app/app.component.ts [hunk 2]
```diff
@@ -1,10 +1,29 @@
...
+import { AuthService } from './services/auth.service';
 
 @Component({
   selector: 'app-root',
   templateUrl: './app.component.html',
   styleUrls: ['./app.component.scss']
 })
 export class AppComponent {
-  constructor() {}
+  str: string;
```
Comment (Duveshka) [inline ArcadiaParties-SPA/src/app/app.component.ts]: Where do you use it?

Comment (Duveshka) [inline ArcadiaParties-SPA/src/app/app.component.ts]: Where do you use it?

Comment (laska-sl) [reply ArcadiaParties-SPA/src/app/app.component.ts]: It was temporary. Already removed.

File: ArcadiaParties-SPA/src/app/app.component.ts (hunk 1 above)
Comment (laska-sl) [reply ArcadiaParties-SPA/src/app/app.component.ts]: Fixed

File: ArcadiaParties-SPA/src/app/services/token-interceptor.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import { HttpRequest, HttpInterceptor, HttpHandler, HttpEvent, HTTP_INTERCEPTORS } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+import { Observable } from 'rxjs';
+
+import { AuthService } from './auth.service';
+import { mergeMap } from 'rxjs/operators';
```
Comment (Duveshka) [inline ArcadiaParties-SPA/src/app/services/token-interceptor.ts]: External and internal

File: ArcadiaParties-SPA/src/app/services/token-interceptor.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+import { AuthService } from './auth.service';
+
+@Injectable()
+export class TokenInterceptor implements HttpInterceptor {
+  constructor(private authService: AuthService) {}
+
+  intercept(request: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
+    return this.authService.acquireTokenResilient().pipe(
```
Comment (sponkrashin) [inline ArcadiaParties-SPA/src/app/services/token-interceptor.ts]: And what happens if you need to login inside? I suppose, the request will never be sent.

File: ArcadiaParties-SPA/src/app/services/unauthorized-interceptor.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+@Injectable()
+export class UnauthorizedInterceptor implements HttpInterceptor {
+  constructor(private authService: AuthService) {}
+
+  intercept(request: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
+    return next.handle(request).pipe(
+      tap(
+        (event: HttpEvent<any>) => {
+          if (event instanceof HttpResponse) {
+          }
```
Comment (sponkrashin) [inline ArcadiaParties-SPA/src/app/services/unauthorized-interceptor.ts]: So what?

File: ArcadiaParties-SPA/src/app/services/auth.service.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    return this.context.getCachedToken('7550c576-a91a-4bd6-8f47-5af6dc701a40');
+  }
+
+  acquireTokenResilient(): Observable<any> {
+    return new Observable<any>((subscriber: Subscriber<any>) =>
+      this.context.acquireToken('7550c576-a91a-4bd6-8f47-5af6dc701a40', (message: string, token: string) => {
+        if (token) {
+          subscriber.next(token);
+        } else {
+          this.login();
```
Comment (sponkrashin) [inline ArcadiaParties-SPA/src/app/services/auth.service.ts]: So you should complete an observable of login is required. Otherwise you'll get hanging observable that is not complete and is not emitting data.

**Review by sponkrashin — CHANGES_REQUESTED** [review]

File: ArcadiaParties-SPA/src/app/services/token-interceptor.ts (hunk 3 above)
Comment (laska-sl) [reply ArcadiaParties-SPA/src/app/services/token-interceptor.ts]: Fixed

File: ArcadiaParties-SPA/src/app/services/unauthorized-interceptor.ts (hunk 5 above)
Comment (laska-sl) [reply ArcadiaParties-SPA/src/app/services/unauthorized-interceptor.ts]: Nothing. Replaced as null optional param

File: ArcadiaParties-SPA/src/app/services/token-interceptor.ts (hunk 4 above)
Comment (laska-sl) [reply ArcadiaParties-SPA/src/app/services/token-interceptor.ts]: Fixed according to your advice

File: ArcadiaParties-SPA/src/app/services/auth.service.ts (hunk 6 above)
Comment (laska-sl) [reply ArcadiaParties-SPA/src/app/services/auth.service.ts]: Fixed according to your advice

File: ArcadiaParties-SPA/src/app/services/auth.service.ts [hunk 7]
```diff
@@ -38,11 +37,7 @@ export class AuthService {
   acquireTokenResilient(): Observable<any> {
     return new Observable<any>((subscriber: Subscriber<any>) =>
       this.context.acquireToken('7550c576-a91a-4bd6-8f47-5af6dc701a40', (message: string, token: string) => {
```
Comment (sponkrashin) [inline ArcadiaParties-SPA/src/app/services/auth.service.ts]: `message` parameter is not used. In such cases it is recommended to use `_` instead of meaning variable name to let TS know that all code style rules shouldn't be applied.

File: ArcadiaParties-SPA/src/app/services/unauthorized-interceptor.ts [hunk 8]
```diff
@@ -15,20 +15,15 @@ import { AuthService } from './auth.service';
...
+  constructor(private authService: AuthService) { }
 
   intercept(request: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
     return next.handle(request).pipe(
       tap(
-        (event: HttpEvent<any>) => {
-          if (event instanceof HttpResponse) {
-          }
-        },
+        null,
```
Comment (sponkrashin) [inline ArcadiaParties-SPA/src/app/services/unauthorized-interceptor.ts]: Use `() => {}` instead.

Comment (sponkrashin) [reply ArcadiaParties-SPA/src/app/services/unauthorized-interceptor.ts]: Otherwise, you could use `catchError` pipe method to catch all errors.

**Review by sponkrashin — CHANGES_REQUESTED** [review]

File: ArcadiaParties-SPA/src/app/services/auth.service.ts (hunk 7 above)
Comment (laska-sl) [reply ArcadiaParties-SPA/src/app/services/auth.service.ts]: I've already done it before your comment, but confused branches to push. Already fixed

File: ArcadiaParties-SPA/src/app/services/unauthorized-interceptor.ts (hunk 8 above)
Comment (laska-sl) [reply ArcadiaParties-SPA/src/app/services/unauthorized-interceptor.ts]: You're right. `tap` is deprecated. I've implemented `catchError`

**Review by sponkrashin — APPROVED** [review]

**Review by sponkrashin — CHANGES_REQUESTED** [review]
Move all authentication parameters to separate json file.

File: ArcadiaParties-SPA/src/app/app-settings/app-settings.service.ts [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+@Injectable({
+  providedIn: 'root'
+})
+export class AppSettingsService {
+  public appSettings: AppSettings;
+
+  constructor(private http: HttpClient) { }
+
+  async init() {
+    console.log();
```
Comment (sponkrashin) [inline ArcadiaParties-SPA/src/app/app-settings/app-settings.service.ts]: Remove it.

File: ArcadiaParties-SPA/src/app/app.module.ts [hunk 10]
```diff
@@ -18,7 +18,11 @@ import { UserModule } from './user/user.module';
 import { TitleEffect } from './effects/effect';
 import { DepartmentModule } from './department/department.module';
 import { TokenInterceptorProvider } from './services/token-interceptor';
-import { UnauthorizedInterceptorProvider } from './services/unauthorized-interceptor';
+import { AppSettingsService } from './app-settings/app-settings.service';
+
+export function initializeAppSettings(appSettingsService: AppSettingsService) {
```
Comment (sponkrashin) [inline ArcadiaParties-SPA/src/app/app.module.ts]: It doesn't need to be exported.

Comment (sponkrashin) [reply ArcadiaParties-SPA/src/app/app.module.ts]: Also it can be inlined, because it is quite simple.

File: ArcadiaParties-SPA/src/app/services/token-interceptor.ts [hunk 11]
```diff
@@ -1,27 +1,41 @@
...
   intercept(request: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
-    return this.authService.acquireTokenResilient().pipe(
-      mergeMap(token => {
-        if (token) {
-          request = request.clone({
-            setHeaders: {
-              Authorization: 'Bearer ' + token
+    if (request.url.includes('/api/')) {
+      return this.injector
+        .get(AuthService)
```
Comment (sponkrashin) [inline ArcadiaParties-SPA/src/app/services/token-interceptor.ts]: It can be get once when first request comes and stored in a variable.

File: ArcadiaParties-SPA/src/app/services/unauthorized-interceptor.ts [hunk 12]
```diff
@@ -1,35 +0,0 @@
-import { Injectable } from '@angular/core';
```
Comment (sponkrashin) [inline ArcadiaParties-SPA/src/app/services/unauthorized-interceptor.ts]: Why did you remove it?

File: ArcadiaParties-SPA/tsconfig.json [hunk 13]
```diff
@@ -1,6 +1,7 @@
 {
   "compileOnSave": false,
   "compilerOptions": {
+    "resolveJsonModule": true,
```
Comment (sponkrashin) [inline ArcadiaParties-SPA/tsconfig.json]: Not needed.

**Review by sponkrashin — CHANGES_REQUESTED** [review]

File: ArcadiaParties-SPA/src/app/services/unauthorized-interceptor.ts (hunk 12 above)
Comment (laska-sl) [reply ArcadiaParties-SPA/src/app/services/unauthorized-interceptor.ts]: We can catch errors in the same interceptor, where we add header

Comment (laska-sl) [reply ArcadiaParties-SPA/src/app/services/unauthorized-interceptor.ts]: We can catch 401 error in the same interceptor, where we add token to header. And it works fine.

File: ArcadiaParties-SPA/src/app/app.module.ts (hunk 10 above)
Comment (laska-sl) [reply ArcadiaParties-SPA/src/app/app.module.ts]: Inlined.

File: ArcadiaParties-SPA/src/app/services/token-interceptor.ts (hunk 11 above)
Comment (laska-sl) [reply ArcadiaParties-SPA/src/app/services/token-interceptor.ts]: Fixed.

File: ArcadiaParties-SPA/tsconfig.json (hunk 13 above)
Comment (laska-sl) [reply ArcadiaParties-SPA/tsconfig.json]: Removed.

File: ArcadiaParties-SPA/src/app/app-settings/app-settings.service.ts (hunk 9 above)
Comment (laska-sl) [reply ArcadiaParties-SPA/src/app/app-settings/app-settings.service.ts]: Removed.

File: ArcadiaParties-SPA/src/app/app.module.ts [hunk 14]
```diff
@@ -20,7 +20,7 @@ import { DepartmentModule } from './department/department.module';
 import { TokenInterceptorProvider } from './services/token-interceptor';
 import { AppSettingsService } from './app-settings/app-settings.service';
 
-export function initializeAppSettings(appSettingsService: AppSettingsService) {
+function initializeAppSettings(appSettingsService: AppSettingsService) {
```
Comment (sponkrashin) [inline ArcadiaParties-SPA/src/app/app.module.ts]: Remove it.

File: ArcadiaParties-SPA/src/app/services/token-interceptor.ts [hunk 15]
```diff
@@ -11,8 +11,9 @@ export class TokenInterceptor implements HttpInterceptor {
 
   intercept(request: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
     if (request.url.includes('/api/')) {
-      return this.injector
-        .get(AuthService)
+      const authService: AuthService = this.injector.get(AuthService);
```
Comment (sponkrashin) [inline ArcadiaParties-SPA/src/app/services/token-interceptor.ts]: Do it only once, not for each request.

**Review by sponkrashin — CHANGES_REQUESTED** [review]

File: ArcadiaParties-SPA/src/app/app.module.ts (hunk 14 above)
Comment (laska-sl) [reply ArcadiaParties-SPA/src/app/app.module.ts]: Removed

File: ArcadiaParties-SPA/src/app/services/token-interceptor.ts (hunk 15 above)
Comment (laska-sl) [reply ArcadiaParties-SPA/src/app/services/token-interceptor.ts]: Done

File: ArcadiaParties-SPA/src/app/services/token-interceptor.ts [hunk 16]
```diff
@@ -7,13 +7,15 @@ import { AuthService } from './auth.service';
 
 @Injectable()
 export class TokenInterceptor implements HttpInterceptor {
+  authService: AuthService;
```
Comment (sponkrashin) [inline ArcadiaParties-SPA/src/app/services/token-interceptor.ts]: `private`.

Comment (sponkrashin) [reply ArcadiaParties-SPA/src/app/services/token-interceptor.ts]: And empty line.

File: ArcadiaParties-SPA/src/app/services/token-interceptor.ts [hunk 17]
```diff
@@ -7,13 +7,15 @@ import { AuthService } from './auth.service';
...
 
   intercept(request: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
     if (request.url.includes('/api/')) {
-      const authService: AuthService = this.injector.get(AuthService);
-
-      return authService
+      if (!this.authService) {
+        this.authService = this.injector.get(AuthService);
+      }
+      return this.authService
```
Comment (sponkrashin) [inline ArcadiaParties-SPA/src/app/services/token-interceptor.ts:20]: Empty line.

**Review by sponkrashin — CHANGES_REQUESTED** [review]

File: ArcadiaParties-SPA/src/app/services/token-interceptor.ts (hunk 16 above)
Comment (laska-sl) [reply ArcadiaParties-SPA/src/app/services/token-interceptor.ts]: Both done

File: ArcadiaParties-SPA/src/app/services/token-interceptor.ts (hunk 17 above)
Comment (laska-sl) [reply ArcadiaParties-SPA/src/app/services/token-interceptor.ts:20]: And this done

**Review by sponkrashin — APPROVED** [review]
